### ISSUE TITLE
[Feature request] In-place versions of some functions, such as cat or padding

### ISSUE THREAD

**Issue by Randl** [pytorch/pytorch#7983]
I propose to introduce in-place versions of some functions, which I believe are often used in a way that can benefit from it. In my case those are `cat` and padding but there are probably more (for example https://github.com/pytorch/pytorch/issues/1641). 

The idea is to make input of the function to share storage with output, for example:
```python3
c = torch.cat_((a,b)) # a is now c[:a_size] and b is c[a_size:]
```
This wouldn't work in all cases, obviously. If there are tensors sharing storage with `a` or `b`, they will be broken. However, I think often we just concat two outputs to pass them forward or pad input to work with padded one, in which cases that would allow to reduce memory usage significantly.
Consider the following example:
```python3
padder = nn.ReflectionPad2d(1)
conv = nn.Conv2d(input_filters, out_filters, kernel_size=3)

padded_input = padder(inp) 
conv_out = F.tanh(conv(padded_input))
output = torch.cat(conv_out, inp)
```
In this case there is no reason to keep separate storage for each of `input`, `padded_input`, and `conv_out` while all of them are actually stored inside the `output`.

This optimizations can probably be performed manually, but this is much less convinient and adds unnecessary complexity to the code

**Comment by soumith** [comment]
I am rejecting this proposal because the proposed optimizations are auto-magical. Such forward-looking optimizations belong to the JIT-compiler, and this should be proposed as a [pass](https://github.com/pytorch/pytorch/tree/master/torch/csrc/jit/passes) for the jit.

Let me elaborate a little more.

In the simple case of

```
c = torch.cat_((a,b)) # a is now c[:a_size] and b is c[a_size:]
```

Initially, `a` and `b` are pointing to separate memory regions. But after the optimization, we allocate a region for `c`, and point `a` and `b` to this memory region, then we deallocate the original memory regions that `a` and `b` point to.

It seems to look reasonable if `a` and `b` are newly allocated and nothing is pointing to them. However, take the following situation:

```
a = A[:, 1]
c = torch.cat_((a,b))
```

in this case, the torch.cat is in-place not just mutating the values of `a` and `b` but actually changing the sharing situation of `a`. `a` is now sharing memory with `c`, but not with `A` anymore.

This situation of magically manipulating storage leads to lots of silent bugs -- in fact we had nightmares with such a situation back in the LuaTorch days when the `getParameters()` function would manipulate `weight` storages automagically.

One can do the same optimization manually (as you pointed out), and that should really be the way forward.

For example:

```
out = torch.empty(....) # create with the size equivalent to size(a)  + size(b)
out[...] = a
out[...] = b
del a, b
```

If it's a common pattern that involves boiler-plate code, for example ReflectionPadding+Conv+Tanh+Identity, make a class for such a layer to reduce the boilerplate
